### ISSUE TITLE
Issues with handling constant non-splat tensors in shardy

### ISSUE THREAD

**Issue by umalesTT** [tenstorrent/tt-mlir#6885]
Currently, some multichip workloads fail with
``` Shardy automatic parallelization currently does not support non-splat constant tensors. ```
in ``` UpdateGlobalToLocalShapes ```, when we try to shard constant tensor that has different values across elements. 

Tensors for which this happens are small and possible fix is to just keep all constant tensor replicated.
This issue happened after [this fix](https://github.com/tenstorrent/tt-mlir/issues/6756)

**Comment by vladimirjovanovicTT** [comment]
Assigning P0 because this is a customer issue - blocks finetuning runs by Evol labs

**Comment by ndrakulicTT** [comment]
For the repro:
1. go to tt-xla
2. checkout `ndrakulic/non_splat` (small change in compiler options for training)
3. run the test `pytest -svv tests/runner/test_models.py::test_all_models_torch[qwen_3/causal_lm/pytorch-0_6b-tensor_parallel-training]`

Workaround for this is to increase num of batches to be grater then 2 per device ( change in `third_party/tt_forge_models/qwen_3/causal_lm/pytorch/loader.py` `def load_inputs(self, dtype_override=None, batch_size=X):`

When batch_size is 4 then it fails with some OOM in runtime (this should be unreleated)

**Comment by sshonTT** [comment]
@umalesTT, @ndrakulicTT 
I've opened a PR [here](https://github.com/tenstorrent/tt-mlir/pull/6928).

With this change, qwen_3/causal_lm/pytorch-0_6b-tensor_parallel-training runs end-to-end, although there is a PCC failure.

Note on PCC Failure, I verified that the PCC error persists even when forcing constant ops to be replicated (bypassing the logic in this patch). This suggests the issue is unrelated to this PR.

However, could you please verify this on your end to confirm?

**Comment by umalesTT** [comment]
@sshonTT 
thanks for the quick fix!
This does fix significant scope of problems with non-splat tensors, but unfortunately not everything - some of our training workloads still fail on this issue, but now with non-splat and non-periodic constant tensor, e.g. ```[0, 1, ..., 63]``` - where shardy tries to shard them and then fails.

Underlying problem with shardy is in their [UserPriorityPropagationPass](https://github.com/openxla/shardy/blob/afa0b5b40aae27ed7c7de7ce46fa5773fec50f7e/shardy/dialect/sdy/transforms/propagation/user_priority_propagation.cc) where they work sort of backwards from consumer ops that consume also inputs (that have defined shardings). My guess is that in uplifted version of transformers, in new attention mask calculation, there is an op which consumes sharded input and non-splat & non-periodic constant which results in sharding of such constant.
Sharding of such constant is logically incorrect (e.g. in positional embeddings per device, we would have offset in positions, as opposed to correct scenario where positions start from 0). 
I have a [small repro](https://gist.github.com/umalesTT/b4abe5eecd9880702d7723fd0b29807d) where such sharding happens and in the IR you will be able to see such sharding on ```%cst_0``` after ```UserPriorityPropagationPass```.
``` IR Dump After UserPriorityPropagationPass (sdy-user-priority-propagate) ('builtin.module' operation: @SyncTensorsGraph.18) 
module @SyncTensorsGraph.18 attributes {mhlo.cross_program_prefetches = [], mhlo.input_output_alias = [], mhlo.is_dynamic = false, mhlo.use_auto_spmd_partitioning = false} {
  sdy.mesh @mesh = <["_axis_0"=8, "_axis_1"=4]>
  func.func @main(%arg0: tensor<64x64xf32> {sdy.sharding = #sdy.sharding<@mesh, [{"_axis_0"}, {}]>, ttcore.argument_type = #ttcore.argument_type<input>, ttcore.runtime_tensor_sharding = #ttcore<runtime_tensor_sharding shard_status = <presharded>, local_shape = tensor<8x64xf32>>}) -> (tensor<f32> {ttcore.runtime_tensor_sharding = #ttcore<runtime_tensor_sharding shard_status = <unsharded>, local_shape = tensor<f32>>}) {
    %cst = stablehlo.constant dense<0.000000e+00> : tensor<f32>
    %cst_0 = stablehlo.constant {sdy.sharding = #sdy.sharding_per_value<[<@mesh, [{"_axis_0", ?}]>]>} dense<[0.000000e+00, 1.000000e+00, 2.000000e+00, 3.000000e+00, 4.000000e+00, 5.000000e+00, 6.000000e+00, 7.000000e+00, 8.000000e+00, 9.000000e+00, 1.000000e+01, 1.100000e+01, 1.200000e+01, 1.300000e+01, 1.400000e+01, 1.500000e+01, 1.600000e+01, 1.700000e+01, 1.800000e+01, 1.900000e+01, 2.000000e+01, 2.100000e+01, 2.200000e+01, 2.300000e+01, 2.400000e+01, 2.500000e+01, 2.600000e+01, 2.700000e+01, 2.800000e+01, 2.900000e+01, 3.000000e+01, 3.100000e+01, 3.200000e+01, 3.300000e+01, 3.400000e+01, 3.500000e+01, 3.600000e+01, 3.700000e+01, 3.800000e+01, 3.900000e+01, 4.000000e+01, 4.100000e+01, 4.200000e+01, 4.300000e+01, 4.400000e+01, 4.500000e+01, 4.600000e+01, 4.700000e+01, 4.800000e+01, 4.900000e+01, 5.000000e+01, 5.100000e+01, 5.200000e+01, 5.300000e+01, 5.400000e+01, 5.500000e+01, 5.600000e+01, 5.700000e+01, 5.800000e+01, 5.900000e+01, 6.000000e+01, 6.100000e+01, 6.200000e+01, 6.300000e+01]> : tensor<64xf32>
    %0 = stablehlo.broadcast_in_dim %cst_0, dims = [0] {sdy.sharding = #sdy.sharding_per_value<[<@mesh, [{"_axis_0", ?}, {?}]>]>} : (tensor<64xf32>) -> tensor<64x64xf32>
    %1 = stablehlo.add %arg0, %0 {sdy.sharding = #sdy.sharding_per_value<[<@mesh, [{"_axis_0", ?}, {?}]>]>} : tensor<64x64xf32>
    %2 = stablehlo.reduce(%1 init: %cst) applies stablehlo.add across dimensions = [0, 1] : (tensor<64x64xf32>, tensor<f32>) -> tensor<f32>
    return %2 : tensor<f32>
  }
}
```
Correct way, in general, would be to keep such constant replicated all the time.

This is the actual core problem that should be addressed and possibly upstreamed to shardy.

I tried quickly fixing this in our stack, and on top of your branch, I added ```ReplicateNonSplittableConstantsPass``` which is right before ```InsertExplicitReshards``` pass (see [here](https://github.com/tenstorrent/tt-mlir/compare/main...umales/pass_for_const)). In this pass, if constant tensor is either splat or periodic (using your logic) we omit them and in other case replication is forced. This is more of a workaround / proof where the problem lies than the actual solution.
With this change, all multichip workloads that I tried passed. 

What are your thoughts on this?

**Comment by sshonTT** [comment]
@umalesTT Yeah, my PR was primarily focused on resolving the periodic constant case. Regarding forcing replication, I actually ran into the same issue. It was non-trivial because the propagation logic tends to override any sharding rules defined on the constant tensor itself. I found that I had to redefine the sharding rules of the consumer ops (like selectOp in my case) to make it work, which wasn't ideal.

While I completely agree that non-splat, non-periodic constants must be replicated, I'm a bit hesitant to introduce a standalone pass just for this specific case. Instead, do you think we could dig a bit deeper into Shardy, perhaps by adjusting the propagation logic or setting stronger constraints to solve this at the root level?

**Comment by umalesTT** [comment]
@sshonTT yeah, we should definitely fix this in Shardy.
The solution with additional pass on our side, as I said in the comment, should only serve as a potential workaround if we run into some complications with real fix on Shardy level :)

**Comment by umalesTT** [comment]
Hey @sshonTT I saw that the PR for periodic tensors was merged. I just wanted to check our plans concerning shardy fix (this is still P0 for training, because in our case tensor is neither splat nor periodic)? 
Thanks!

**Comment by umalesTT** [comment]
I put up a [PR](https://github.com/tenstorrent/tt-mlir/pull/7020) for this. 
Solving this issue at a root level in Shardy seems pretty complex; all discussed approaches reduce to case by case work depending on consumer ops. 
I was unable to come up with anything better, I'd be happy to hear and compare possibly better ideas, if there are any.
